## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [127]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [129]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [131]:
# Use the describe function to gather some basic statistics
print(df.describe())

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [133]:
# Use this space to do any additional research and familiarize yourself with the data.

# Check for missing values in each column
print(df.isnull().sum())

first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64


In [155]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().nlargest(3)
print("Top three item categories with most entries:")
print(top_categories)

Top three item categories with most entries:
category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [135]:
# Check for outliers in numerical columns:
# You can use box plots or z-scores to identify outliers
# For box plots:
df.boxplot(column=['unit_price', 'unit_cost', 'unit_weight', 'qty'])
# For z-scores:
from scipy import stats
z_scores = stats.zscore(df[['unit_price', 'unit_cost', 'unit_weight', 'qty']])
print(z_scores)

       unit_price  unit_cost  unit_weight       qty
0        5.223937   4.980841     0.468574 -0.024754
1       -0.605408  -0.633480    -0.659736 -0.029223
2       -0.667571  -0.687774    -0.624066 -0.028266
3       -0.543027  -0.560186    -0.708548 -0.028798
4        0.319967   0.065513     7.777244 -0.029277
...           ...        ...          ...       ...
54634   -0.288991  -0.359304    -0.517054 -0.028585
54635    0.382456   0.570833     1.257077 -0.027840
54636   -0.384003  -0.316425    -0.158473 -0.005069
54637   -0.733051  -0.737638     2.447341 -0.024382
54638   -0.724730  -0.734709    -0.550847  0.024513

[54639 rows x 4 columns]


In [137]:
# Check the data types of each column:
print(df.dtypes)

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object


In [139]:
# View unique values in the 'category' column
print(df['category'].unique())

['decor' 'consumables' 'software' 'furniture' 'electronics']


In [141]:
#Check for duplicate rows:
print(df.duplicated().sum())

0


In [145]:
# For the category with the most entries, which subcategory had the most entries?
# Identify the category with the most entries
most_common_category = top_categories.index[0]

# Filter the dataframe to only include rows with the most common category
category_df = df[df['category'] == most_common_category]

# Find the subcategory with the most entries within the most common category
top_subcategory = category_df['subcategory'].value_counts().idxmax()
print(top_subcategory)

bathroom supplies


In [159]:
# For the category with the most entries, which subcategory had the most entries?
most_entries_category = top_categories.index[0]
top_subcategory = df[df['category'] == most_entries_category]['subcategory'].value_counts().idxmax()
print("Subcategory with most entries in the category with most entries:", top_subcategory)


Subcategory with most entries in the category with most entries: bathroom supplies


In [147]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
print(top_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [149]:
# Store the client ids of those top 5 clients in a list.
top_clients_list = top_clients.index.tolist()
print(top_clients_list)

[33615, 66037, 46820, 38378, 24741]


In [151]:
# How many total units (the qty column) did the client with the most entries order order?
# Identify the client with the most entries
top_client_id = top_clients.index[0]

# Filter the dataframe to only include rows with the top client ID
client_df = df[df['client_id'] == top_client_id]

# Calculate the total units ordered by the client with the most entries
total_units_ordered = client_df['qty'].sum()
print(total_units_ordered)

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [59]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
print(df['subtotal'])
print(df['qty'])

0        115164.00
1           523.95
2           527.28
3          1056.18
4          3902.00
           ...    
54634      2743.29
54635      9709.73
54636     31188.50
54637       165.76
54638      3103.31
Name: subtotal, Length: 54639, dtype: float64
0         105
1          21
2          39
3          29
4          20
         ... 
54634      33
54635      47
54636     475
54637     112
54638    1031
Name: qty, Length: 54639, dtype: int64


In [63]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
# Create a column for the shipping price
df['shipping_price'] = df['unit_weight'].apply(lambda x: 7*x if x > 50 else 10*x)
print(df['shipping_price'])
print(df['unit_weight'])

0         75.0
1         14.9
2         16.8
3         12.3
4        464.3
         ...  
54634     22.5
54635    117.0
54636     41.6
54637    180.4
54638     20.7
Name: shipping_price, Length: 54639, dtype: float64
0         7.50
1         1.49
2         1.68
3         1.23
4        46.43
         ...  
54634     2.25
54635    11.70
54636     4.16
54637    18.04
54638     2.07
Name: unit_weight, Length: 54639, dtype: float64


In [65]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925
print(df['total_price'])

0        125898.607500
1           588.693625
2           594.407400
3          1167.314400
4          4770.182750
             ...      
54634      3021.625575
54635     10735.702525
54636     34118.884250
54637       378.179800
54638      3412.980925
Name: total_price, Length: 54639, dtype: float64


In [69]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
print(df['line_cost'])
print(df['shipping_price'])

0        80159.55
1          331.79
2          323.34
3          732.95
4         2627.70
           ...   
54634     1725.30
54635     8363.62
54636    27263.85
54637      317.04
54638     1680.61
Name: line_cost, Length: 54639, dtype: float64
0         75.0
1         14.9
2         16.8
3         12.3
4        464.3
         ...  
54634     22.5
54635    117.0
54636     41.6
54637    180.4
54638     20.7
Name: shipping_price, Length: 54639, dtype: float64


In [71]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['total_price'] - df['line_cost']
print(df['profit'])
print(df['total_price'])
print(df['line_cost'])

0        45739.057500
1          256.903625
2          271.067400
3          434.364400
4         2142.482750
             ...     
54634     1296.325575
54635     2372.082525
54636     6855.034250
54637       61.139800
54638     1732.370925
Name: profit, Length: 54639, dtype: float64
0        125898.607500
1           588.693625
2           594.407400
3          1167.314400
4          4770.182750
             ...      
54634      3021.625575
54635     10735.702525
54636     34118.884250
54637       378.179800
54638      3412.980925
Name: total_price, Length: 54639, dtype: float64
0        80159.55
1          331.79
2          323.34
3          732.95
4         2627.70
           ...   
54634     1725.30
54635     8363.62
54636    27263.85
54637      317.04
54638     1680.61
Name: line_cost, Length: 54639, dtype: float64


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [42]:
# Check your work using the totals above
# Verify total prices for specific order IDs
order_2742071_total = df[df['order_id'] == 2742071]['total_price'].sum()
order_2173913_total = df[df['order_id'] == 2173913]['total_price'].sum()
order_6128929_total = df[df['order_id'] == 6128929]['total_price'].sum()

print(f"Order 2742071 Total: {order_2742071_total}")
print(f"Order 2173913 Total: {order_2173913_total}")
print(f"Order 6128929 Total: {order_6128929_total}")


Order 2742071 Total: 139844.51202499997
Order 2173913 Total: 119838.35704999999
Order 6128929 Total: 752748.7454750001


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [44]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
# Calculate the total spending for each of the top 5 clients
top_clients_spending = df[df['client_id'].isin(top_clients_list)].groupby('client_id')['total_price'].sum()
print(top_clients_spending)


client_id
24741    7.667848e+07
33615    6.392918e+06
38378    9.172866e+06
46820    8.006333e+06
66037    8.746255e+06
Name: total_price, dtype: float64


In [46]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
# Create a summary DataFrame
summary_df = df[df['client_id'].isin(top_clients_list)].groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('profit', 'sum')
).reset_index()

print(summary_df)

   client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      24741                 239862                9365.6   7.667848e+07   
1      33615                  64313               12609.4   6.392918e+06   
2      38378                  73667               11895.0   9.172866e+06   
3      46820                  75768               11094.8   8.006333e+06   
4      66037                  43018               10017.3   8.746255e+06   

   total_profit  
0  3.610666e+07  
1  2.033980e+06  
2  2.955706e+06  
3  2.589495e+06  
4  3.126907e+06  


In [87]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns.
money_columns = ['total_shipping_price', 'total_revenue', 'total_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    return f'${x/1e6:.2f}M'

# Check the existing columns in summary_df
print("Existing columns in summary_df:", summary_df.columns)

# Apply the currency_format_millions function to only the money columns if they exist.
for col in money_columns:
    if col in summary_df.columns:
        summary_df[col] = summary_df[col].apply(currency_format_millions)
    else:
        print(f"Warning: Column '{col}' not found in summary_df")

# Rename the columns to reflect the change in the money format.
summary_df = summary_df.rename(columns={
    'total_units_purchased': 'Total Units Purchased',
    'total_shipping_price': 'Total Shipping Price (millions)',
    'total_revenue': 'Total Revenue (millions)',
    'total_profit': 'Total Profit (millions)'
})

print(summary_df)


Existing columns in summary_df: Index(['client_id', 'Total Units Purchased', 'Total Shipping Price (millions)',
       'Total Revenue (millions)', 'Total Profit (millions)'],
      dtype='object')
   client_id  Total Units Purchased Total Shipping Price (millions)  \
0      24741                 239862                          $0.01M   
1      33615                  64313                          $0.01M   
2      38378                  73667                          $0.01M   
3      46820                  75768                          $0.01M   
4      66037                  43018                          $0.01M   

  Total Revenue (millions) Total Profit (millions)  
0                  $76.68M                 $36.11M  
1                   $6.39M                  $2.03M  
2                   $9.17M                  $2.96M  
3                   $8.01M                  $2.59M  
4                   $8.75M                  $3.13M  


In [175]:
# Format the data and rename the columns
def currency_format_millions(x):
    return '${:,.2f}M'.format(x/1000000)

money_columns = ['total_shipping_price', 'total_revenue', 'total_profit']
summary_df[money_columns] = summary_df[money_columns].applymap(currency_format_millions)
summary_df = summary_df.rename(columns={
    'total_units': 'Total Units Purchased',
    'total_shipping_price': 'Total Shipping Price',
    'total_revenue': 'Total Revenue',
    'total_profit': 'Total Profit (millions)'
})

KeyError: "None of [Index(['total_shipping_price', 'total_revenue', 'total_profit'], dtype='object')] are in the [columns]"

In [79]:
# Sort the updated data by "Total Profit (millions)" from highest to lowest and assign the sort to a new DataFrame.
sorted_summary_df = summary_df.sort_values(by='Total Profit (millions)', ascending=False).reset_index(drop=True)

print(sorted_summary_df)

   client_id  Total Units Purchased Total Shipping Price (millions)  \
0      24741                 239862                          $0.01M   
1      66037                  43018                          $0.01M   
2      38378                  73667                          $0.01M   
3      46820                  75768                          $0.01M   
4      33615                  64313                          $0.01M   

  Total Revenue (millions) Total Profit (millions)  
0                  $76.68M                 $36.11M  
1                   $8.75M                  $3.13M  
2                   $9.17M                  $2.96M  
3                   $8.01M                  $2.59M  
4                   $6.39M                  $2.03M  


Summary of findings:
# The top 5 clients by quantity ordered a significant amount of units.
# The total revenue generated by these clients is substantial, with noticeable profits.
# The total shipping costs are considerable but necessary for the revenue generated.